### PR TITLE
add section about async hooks

### LINKED ISSUE #136: update parser hooks description regarding async handlers
See also https://github.com/xtermjs/xterm.js/pull/3222.

### PR DESCRIPTION
**Pull request by jerch**
Closes #136.

### CODE REVIEW

File: _docs/guides/hooks.md [hunk 1]
```diff
@@ -221,10 +222,74 @@ Hopefully this small example illustrates the power of the hooks system. It makes
...
+    .then(more)
+    .then(steps)
+    .then(() => false)
+});
+```
+Or with `async`/`await` syntax (discouraged if there is a synchronous side path):
+
+```typescript
+const myCustomHandler = term.registerCsiHandler(..., async params => {
+  const someWaitingResult: any = await asyncInterface(...);
```
Comment (Tyriar) [inline _docs/guides/hooks.md:250]: Did you mean `Promise<any>`?

File: _docs/guides/hooks.md [hunk 2]
```diff
@@ -221,10 +222,74 @@ Hopefully this small example illustrates the power of the hooks system. It makes
...
+  await steps(await more(someWaitingResult));
+  return false;
+});
+```
+
+Whenever a handler returns a promise, the parser stops input processing at the current band position
+and returns control to the top level (stack unwinding). Input processing will not resume until
+that promise got fullfilled (resolved or rejected). With this in-band blocking the parser can
```
Comment (Tyriar) [inline _docs/guides/hooks.md:258]:
```suggestion
that promise got fulfilled (resolved or rejected). With this in-band blocking the parser can
```

File: _docs/guides/hooks.md [hunk 3]
```diff
@@ -221,10 +222,74 @@ Hopefully this small example illustrates the power of the hooks system. It makes
...
+input data before the active async handler finally finished its work.
+
+
+Caveats regarding async parser handlers:
+- **general slowdown of input processing**  
+  Due to needed stack unwinding async handlers come to a rather high price of poor input throughput.
+  They should only be used if a synchronous implementation is not applicable.
+- **can block terminal input forever**  
+  Make sure that a returned promise will always be fullfilled in a timely fashion to avoid
+  poor user experience. With `logLevel.WARN` or higher the terminal will warn about an async handler
```
Comment (Tyriar) [inline _docs/guides/hooks.md:269]:
```suggestion
  poor user experience. With `logLevel.WARN` or higher, the terminal will warn about an async handler
```

**Review by Tyriar — APPROVED** [review]
Thanks for fixing the deps up

File: _docs/guides/hooks.md (hunk 1 above)
Comment (jerch) [reply _docs/guides/hooks.md:250]: Nope, as it is already awaited on the right side. But I will change the example back to the promise and await it below, to be more in line with the first example..

**Pull request merged**